Starcounter.Uniform/ViewModels/UniDataTable.json.cs

```diff
@@ -0,0 +1,161 @@
+using Starcounter.Uniform.Builder;
+using Starcounter.Uniform.Generic.FilterAndSort;
+using Starcounter.Uniform.Generic.Pagination;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Starcounter.Uniform.ViewModels
+{
+    public partial class UniDataTable : Json
+    {
+        public IFilteredDataProvider<Json> FilteredDataProvider { get; set; }
+
+        public UniDataTable Init(IFilteredDataProvider<Json> dataProvider, IEnumerable<DataTableColumn> sourceColumns, int initialPageSize, int initialPageIndex)
+        {
+            this.FilteredDataProvider = dataProvider;
+
+            Pagination.DataProvider = dataProvider;
+            Pagination.LoadRows = LoadRows;
+
+            this.FilteredDataProvider.PaginationConfiguration = new PaginationConfiguration(100, initialPageIndex);
+            this.FilteredDataProvider.FilterOrderConfiguration = new FilterOrderConfiguration();
+
+            PopulateColumns(sourceColumns);
+            LoadRows();
+
+            return this;
+        }
+
+        private void LoadRows()
+        {
+            var page = this.FilteredDataProvider.PaginationConfiguration.CurrentPageIndex;
+            if (page > 0)
+            {
+                // Add missing dummy pages to maintain sparse page indicies in Pages
+                while (this.Pages.ElementAtOrDefault(page - 1) == null)
+                {
+                    this.Pages.Add();
+                }
+            }
+
+            var newRowsData = new PagesViewModel();
+
+            foreach (var currentPageRow in this.FilteredDataProvider.CurrentPageRows)
+            {
+                newRowsData.Rows.Add(currentPageRow);
+            }
+
+            if (this.Pages.ElementAtOrDefault(page) == null)
+            {
+                this.Pages.Insert(page, newRowsData);
+            }
+            else
+            {
+                this.Pages[page] = newRowsData;
+            }
+
+            this.TotalRows = this.FilteredDataProvider.TotalRows;
+        }
+
+        private void PopulateColumns(IEnumerable<DataTableColumn> sourceColumns)
+        {
+            foreach (var sourceColumn in sourceColumns)
+            {
+                var column = this.Columns.Add();
+                column.Data = sourceColumn;
+                column.DataProvider = this.FilteredDataProvider;
+                column.LoadRows = LoadRows;
+            }
+        }
+
+        [UniDataTable_json.Pagination]
+        public partial class PaginationViewModel : Json
+        {
+            public IFilteredDataProvider<Json> DataProvider { get; set; }
+            public Action LoadRows { get; set; }
+
+            public int PageSize => DataProvider.PaginationConfiguration.PageSize;
+
+            public int PagesCount => (DataProvider.TotalRows + DataProvider.PaginationConfiguration.PageSize - 1) / DataProvider.PaginationConfiguration.PageSize;
+
+            public void Handle(Input.CurrentPageIndex action)
```

Comment:
what happens if invalid page is selected? Is this tested?

Reply:
I'll add test for that. From now on, when given page index is bigger than total pages count, last page index will be set instead. If given page index will be lower than zero, zero will be used as index instead.